packages/rpc-methods/src/restricted/dialog.ts

```diff
@@ -0,0 +1,298 @@
+  methodHooks: DialogMethodHooks;
+};
+
+type DialogSpecification = ValidPermissionSpecification<{
+  permissionType: PermissionType.RestrictedMethod;
+  targetKey: typeof methodName;
+  methodImplementation: ReturnType<typeof getDialogImplementation>;
+  allowedCaveats: Readonly<NonEmptyArray<string>> | null;
+}>;
+
+/**
+ * The specification builder for the `snap_dialog` permission. `snap_dialog`
+ * lets the Snap display one of the following dialogs to the user:
+ * - An alert, for displaying information.
+ * - A confirmation, for accepting or rejecting some action.
+ * - A prompt, for inputting some information.
+ *
+ * @param options - The specification builder options.
+ * @param options.allowedCaveats - The optional allowed caveats for the permission.
+ * @param options.methodHooks - The RPC method hooks needed by the method implementation.
+ * @returns The specification for the `snap_dialog` permission.
+ */
+const specificationBuilder: PermissionSpecificationBuilder<
+  PermissionType.RestrictedMethod,
+  DialogSpecificationBuilderOptions,
+  DialogSpecification
+> = ({
+  allowedCaveats = null,
+  methodHooks,
+}: DialogSpecificationBuilderOptions) => {
+  return {
+    permissionType: PermissionType.RestrictedMethod,
+    targetKey: methodName,
+    allowedCaveats,
+    methodImplementation: getDialogImplementation(methodHooks),
+  };
+};
+
+export const dialogBuilder = Object.freeze({
+  targetKey: methodName,
+  specificationBuilder,
+  methodHooks: {
+    showAlert: true,
+    showConfirmation: true,
+    showPrompt: true,
+  },
+} as const);
+
+type AlertParameters = {
+  type: DialogType.alert;
+  fields: AlertFields;
+};
+
+type ConfirmationParameters = {
+  type: DialogType.confirmation;
+  fields: ConfirmationFields;
+};
+
+type PromptParameters = {
+  type: DialogType.prompt;
+  fields: PromptFields;
+};
+
+export type DialogParameters =
+  | AlertParameters
+  | ConfirmationParameters
+  | PromptParameters;
+
+/**
+ * Builds the method implementation for `snap_dialog`.
+ *
+ * @param hooks - The RPC method hooks.
+ * @param hooks.showAlert - A function that shows an alert in the MetaMask UI
+ * and returns when the user has closed the alert.
+ * @param hooks.showConfirmation - A function that shows a dialog in the
+ * MetaMask UI and returns a `boolean` that signals whether the user
+ * approved or denied the confirmation.
+ * @param hooks.showPrompt - A function that shows a prompt in the MetaMask UI
+ * and returns the value the user entered into the prompt's text field.
+ * @returns The method implementation which returns `true` if the user approved the confirmation, otherwise `false`.
+ */
+export function getDialogImplementation({
+  showAlert,
+  showConfirmation,
+  showPrompt,
+}: DialogMethodHooks) {
+  // This rule appears to trigger because ESLint does not understand execution
+  // will never reach the "end" of this function.
+  // eslint-disable-next-line consistent-return
```

Comment:
Would this be solved by `return assertExhaustive(type);`, or would that result in TypeScript errors?

Reply:
That solves it